src/Microsoft.AspNet.OData.Shared/Extensions/ODataPathExtensions.cs

```diff
@@ -41,6 +41,59 @@ internal static class ODataPathExtensions
             return keys;
         }
 
+        public static ODataPathSegment GetLastNonKeySegment(this ODataPath path)
+        {
+            if (path == null)
+            {
+                throw Error.ArgumentNull(nameof(path));
+            }
+
+            if (path.Count == 1)
+            {
+                return path.LastSegment;
+            }
+
+            if (path.LastSegment is TypeSegment)
+            {
+                ODataPath modifiedPath = path.TrimEndingTypeSegment();
+
+                return modifiedPath.LastSegment;
+            }
+
+            return path.LastSegment;
+        }
+
+        public static ODataPathSegment GetLastNonTypeNonKeySegment(this ODataPath path)
+        {
+            if (path == null)
+            {
+                throw Error.ArgumentNull(nameof(path));
+            }
+
+            if (path.Count == 1)
+            {
+                return path.LastSegment;
+            }
+
+            ODataPath modifiedPath = path;
```

Comment:
why are assigning path to a variable here. can you use the path directly?

Reply:
Fixed